Guides/Change-Streams.md

```diff
@@ -8,158 +8,155 @@ MongoSwift 0.2.0 added support for [change streams](https://docs.mongodb.com/man
+	stream.forEach { event in
+		// process `MyCodableType` here
+	}
+}.whenFailure { error in
+	// handle error
 }
 
-// check if any errors occurred while iterating
-if let error = stream.error {
-    // handle error
-}
+// perform some operations using `inventory`...
 ```
 
 ### Open a Change Stream on a `MongoDatabase` (MongoDB 4.0+)
 ```swift
-let client = try MongoClient()
+let elg = MultiThreadedEventLoopGroup(numberOfThreads: 4)
+let client = try MongoClient(using: elg)
 let db = client.db("example")
-let stream = try db.watch()
 
-// perform some operations using `db`...
-
-for change in stream {
-    // process `ChangeStreamEvent<Document>` here
+db.watch().flatMap { stream in // a `ChangeStream<ChangeStreamEvent<Document>>`
+	stream.forEach { event in
+		// process `ChangeStreamEvent<Document>` here
+	}
+}.whenFailure { error in
+	// handle error
 }
 
-// check if any errors occurred while iterating
-if let error = stream.error {
-    // handle error
-}
+// perform some operations using `db`...
 ```
 
 Note: the types of the `fullDocument` property, as well as the return type of `ChangeStream.next()`, may be customized in the same fashion as the examples using `MongoCollection` above.
 
 ### Open a Change Stream on a `MongoClient` (MongoDB 4.0+)
 ```swift
-let client = try MongoClient()
-let stream = try client.watch()
+let elg = MultiThreadedEventLoopGroup(numberOfThreads: 4)
+let client = try MongoClient(using: elg)
 
-// perform some operations using `client`...
-
-for change in stream {
-    // process `ChangeStreamEvent<Document>` here
+client.watch().flatMap { stream in // a `ChangeStream<ChangeStreamEvent<Document>>`
+	stream.forEach { event in
+		// process `ChangeStreamEvent<Document>` here
+	}
+}.whenFailure { error in
+	// handle error
 }
 
-// check if any errors occurred while iterating
-if let error = stream.error {
-    // handle error
-}
+// perform some operations using `client`...
 ```
 
 Note: the types of the `fullDocument` property, as well as the return type of `ChangeStream.next()`, may be customized in the same fashion as the examples using `MongoCollection` above.
 
 ### Resume a Change Stream
 ```swift
-let client = try MongoClient()
+let elg = MultiThreadedEventLoopGroup(numberOfThreads: 4)
+let client = try MongoClient(using: elg)
 let inventory = client.db("example").collection("inventory")
-let stream = try inventory.watch()
 
-// perform some operations using `inventory`...
-
-// read the first change event
-let next = try stream.nextOrError()
-
-// create a new change stream that starts after the first change event
-let resumeToken = stream.resumeToken
-let resumedStream = try inventory.watch(options: ChangeStreamOptions(resumeAfter: resumeToken))
-for change in resumedStream {
-    // process `ChangeStreamEvent<Document>` here
+inventory.watch().flatMap { stream in // a `ChangeStream<ChangeStreamEvent<Document>>`
```

Comment:
I've done this but not sure if it's the way you were thinking. Let me know what you think!

Reply:
It's really close, but it's a little racy because it doesn't guarantee the resume attempt happens after killing the stream. The `stream.next()` invocation returns immediately and starts some operations in the background, and the next line we check for the session id. If `next` hasn't completed by the time we check for the id, then it will still be `nil`.

To alleviate this, we can just use `flatMap` chains to guarantee they will be ordered (this in one line for demonstration purposes):
```swift
stream.next().flatMap { _ in stream.kill() }.flatMap { /* resume */ }
```
Using a single `flatMap` chain also has the added benefit that the single `whenFailure` will catch all errors. In the existing implementation, we don't have any callbacks handling the errors of `next` or `kill` since we're only returning the future from `inventory.watch`.

Reply:
Got it - I've made the necessary changes!